### PR TITLE
Remove numpy as a required run dependency

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/apache/arrow/issues/44653

We removed numpy as a required dependency in the pyproject.toml listed dependencies (and thus on PyPI), but forgot about bringing that up when pyarrow 18.0 was added to conda-forge.

I am not super familiar with how run_exports and run_constrained work, so this is just a quick attempt, and only for pyarrow-core for now

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -92,6 +93,7 @@ outputs:
         - __cuda >={{ cuda_compiler_version_min }}  # [cuda_compiler_version != "None"]
       run_constrained:
         - apache-arrow-proc =*={{ build_ext }}
+        - numpy
```
Comment (jorisvandenbossche) [inline recipe/meta.yaml:96]: When adding numpy as is here to `run_constrained`, will it still pick up the version range from numpy's `run_export`? Or if listing numpy here, do we need to manually add the version range?

**Comment by h-vetinari** [comment]
> We removed numpy as a required dependency in the pyproject.toml listed dependencies (and thus on PyPI), but forgot about bringing that up when pyarrow 18.0 was added to conda-forge.

Is numpy still required to build or for any of the other pyarrow-* outputs?

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml:96]: If you ignore the run-export, then nothing will be picked up. If you want to impose a range, we'll have to "manually" add the version constraints from the run-export here.

**Comment by jorisvandenbossche** [comment]
Yes, numpy is still a build dependency (so we need to keep it in build/host) for all of the compiled code in pyarrow (but I think the building is only done for `pyarrow-core`?)

File: recipe/meta.yaml (hunk 1 above)
Comment (jorisvandenbossche) [reply recipe/meta.yaml:96]: And doing something like `pin_compatible('numpy')` also won't work? (I see pin_compatible being used in some other recipes in the run_constrained section)

**Comment by h-vetinari** [comment]
> but I think the building is only done for `pyarrow-core`?

No, the building is done once for all outputs, and then we only add the respective `libarrow-*` dependencies where they belong.

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml:96]: `pin_compatible` is just a fancy way of saying "take version `X.Y` of this dependency at build time, and create a runtime constraint `>=X.Y,<{X+1}` (and so on if the max_pin is tighter than major-level)

That's actually not fully accurate for numpy, which [uses](https://github.com/conda-forge/numpy-feedstock/blob/1b3ce6a006b96f3356d252ac874907398a88a910/recipe/meta.yaml#L51)
```
numpy >={{ default_abi_level }},<3
```
where -- very unusually -- the runtime lower bound is allowed to be lower than at build time. For numpy 2.1, it's `>=1.21,<3` for example.

However, using `pin_compatible("numpy")` would still be good enough. It'd be tighter than it needs to be (effectively forcing pyarrow users to upgrade to numpy 2.0, whereas before they were still able to use 1.x), but perhaps that's not a big deal.

Comment (jorisvandenbossche) [reply recipe/meta.yaml:96]: > It'd be tighter than it needs to be (effectively forcing pyarrow users to upgrade to numpy 2.0, whereas before they were still able to use 1.x), but perhaps that's not a big deal.

I think I would prefer not forcing users to upgrade to numpy  >= 2 through an update of pyarrow (we deliberately kept compatibility with 1.x in pyarrow)

**Comment by jorisvandenbossche** [comment]
> No, the building is done once for all outputs,

Yes, but that "building once" is done through the pyarrow-core output? (at least that's what I meant with my previous message; this is not so important, just want to ensure my understanding about how the feedstock works is correct)

**Comment by h-vetinari** [comment]
> Yes, but that "building once" is done through the pyarrow-core output?

That's correct, see
https://github.com/conda-forge/pyarrow-feedstock/blob/73c212efcee0d7a65886f35627fd6c729c61e8c0/recipe/meta.yaml#L29-L31

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml:96]: It should be possible to do something like
```
- {{ pin_compatible("numpy", lower_bound="1.21", max_pin="x") }}
```

**Comment by jorisvandenbossche** [comment]
One question here is if we remove numpy from all conda-forge outputs, or only from `pyarrow-core` ?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -92,6 +93,9 @@ outputs:
         - __cuda >={{ cuda_compiler_version_min }}  # [cuda_compiler_version != "None"]
       run_constrained:
         - apache-arrow-proc =*={{ build_ext }}
+        # keep lower pin aligned with run_exports from numpy
+        # https://github.com/conda-forge/numpy-feedstock/blob/main/recipe/meta.yaml
+        - {{ pin_compatible("numpy", lower_bound="1.21", max_pin="x") }}
```
Comment (h-vetinari) [inline recipe/meta.yaml:98]: Nevermind, conda-build gets this wrong when it crosses more than a major version (the v1 recipe format will do better). We'll have to do it fully manually
```suggestion
        - numpy >=1.21,<3
```

**Comment by h-vetinari** [comment]
> One question here is if we remove numpy from all conda-forge outputs, or only from `pyarrow-core` ?

I think you could remove it from all the outputs except pyarrow-tests.

**Comment by jorisvandenbossche** [comment]
The two failing builds are flight tests that abort (which I am assuming is probably not related to this PR, not sure if that happens regularly)

**Comment by h-vetinari** [comment]
> The two failing builds are flight tests that abort (which I am assuming is probably not related to this PR, not sure if that happens regularly)

Yeah, there are some unfixed segfaults on ppc unfortunately, but they only come and go and the frequency is low enough that I haven't skipped those tests.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -35,6 +35,7 @@ outputs:
       ignore_run_exports_from:
         - {{ compiler("cuda") }}                 # [cuda_compiler_version != "None"]
         - libarrow-all
+        - numpy
```
Comment (h-vetinari) [inline recipe/meta.yaml:38]: Could we write a comment here like
```suggestion
        # we don't need numpy at runtime, just to build
        - numpy
```
(I haven't actually understood what happens in arrow there, so would appreciate you coming up with a formulation).

Comment (jorisvandenbossche) [reply recipe/meta.yaml:39]: That's essentially it yes: it is not required at run-time (you can import pyarrow and use functionality that doesn't require numpy), but we still always build the numpy-related functionality (we are using the numpy C API), so it is still a required build dependency.

**Review by h-vetinari — APPROVED** [review]
Thanks!

**Comment by jorisvandenbossche** [comment]
@h-vetinari thanks for the help!

**Pull request merged**